File: services/ai-coach.ts
import { get, post, put, del } from './api';
import { getToken } from './auth';

export interface AICoachPreferences {
  id: string;
  user_id: string;
  timezone: string;
  language: string;
  daily_digest_enabled: boolean;
  daily_digest_hour: number;
  daily_digest_minute: number;
  weekly_digest_enabled: boolean;
  monthly_digest_enabled: boolean;
  yearly_digest_enabled: boolean;
  critique_style: 'gentle' | 'balanced' | 'strict';
  quiet_hours_enabled: boolean;
  quiet_hours_start?: string | null;
  quiet_hours_end?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface AICoachDigestReport {
  id: string;
  user_id: string;
  digest_type: 'daily' | 'weekly' | 'monthly' | 'yearly';
  period_key: string;
  target_date: string;
  language: string;
  title: string;
  summary: string;
  score?: number;
  delivered: boolean;
  delivered_at?: string | null;
  delivery_channel?: string | null;
  created_by: string;
  created_at?: string;
  parsed?: Record<string, unknown>;
  raw?: string;
}

export interface RunDigestResponse {
  status: 'created' | 'updated' | 'exists' | 'skipped';
  reason?: string;
  report?: AICoachDigestReport;
  push_result?: {
    sent: number;
    failed: number;
    expired: number;
  };
}

export interface AICoachMemoryItem {
  id: string;
  type: string;
  key: string;
  summary: string;
  value?: Record<string, unknown>;
  confidence: number;
  salience: number;
  last_reinforced_at?: string | null;
}

function requireToken(): string {
  const token = getToken();
  if (!token) {
    throw new Error('Authentication required');
  }
  return token;
}

export async function getAICoachPreferences(): Promise<AICoachPreferences> {
  return get<AICoachPreferences>('/ai-coach/preferences', requireToken());
}

export async function updateAICoachPreferences(
  payload: Omit<AICoachPreferences, 'id' | 'user_id' | 'created_at' | 'updated_at'>
): Promise<AICoachPreferences> {
  return put<AICoachPreferences>('/ai-coach/preferences', payload, requireToken());
}

export async function getAICoachReports(
  limit: number = 20,
  digestType?: 'daily' | 'weekly' | 'monthly' | 'yearly'
): Promise<AICoachDigestReport[]> {
  const params = new URLSearchParams({ limit: String(limit) });
  if (digestType) params.set('digest_type', digestType);
  return get<AICoachDigestReport[]>(`/ai-coach/reports?${params.toString()}`, requireToken());
}

export async function runAICoachDigest(payload: {
  digest_type?: 'daily' | 'weekly' | 'monthly' | 'yearly';
  target_date?: string;
  force?: boolean;
  deliver_notification?: boolean;
}): Promise<RunDigestResponse> {
  return post<RunDigestResponse>(
    '/ai-coach/run-digest',
    {
      digest_type: payload.digest_type || 'daily',
      target_date: payload.target_date,
      force: payload.force ?? true,
      deliver_notification: payload.deliver_notification ?? true,
    },
    requireToken()
  );
}

export async function getAICoachMemory(limit: number = 20): Promise<AICoachMemoryItem[]> {
  return get<AICoachMemoryItem[]>(`/ai-coach/memory?limit=${limit}`, requireToken());
}

export async function deleteAICoachMemory(memoryId: string): Promise<void> {
  await del(`/ai-coach/memory/${memoryId}`, undefined, requireToken());
}

File: backend/app/schemas/ai_coach.py
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class AICoachPreferenceBase(BaseModel):
    timezone: str = Field(default="UTC", max_length=64)
    language: str = Field(default="fa", max_length=8)

    daily_digest_enabled: bool = True
    daily_digest_hour: int = Field(default=20, ge=0, le=23)
    daily_digest_minute: int = Field(default=0, ge=0, le=59)

    weekly_digest_enabled: bool = True
    monthly_digest_enabled: bool = True
    yearly_digest_enabled: bool = True

    critique_style: Literal["gentle", "balanced", "strict"] = "balanced"

    quiet_hours_enabled: bool = False
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None


class AICoachPreferenceUpdate(AICoachPreferenceBase):
    pass


class AICoachPreferenceResponse(AICoachPreferenceBase):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class AICoachDigestRunRequest(BaseModel):
    digest_type: Literal["daily", "weekly", "monthly", "yearly"] = "daily"
    target_date: Optional[str] = None  # YYYY-MM-DD
    force: bool = False
    deliver_notification: bool = True


class AICoachDigestReportResponse(BaseModel):
    id: str
    user_id: str
    digest_type: str
    period_key: str
    target_date: str
    language: str
    title: str
    summary: str
    score: Optional[float] = None
    delivered: bool
    delivered_at: Optional[str] = None
    delivery_channel: Optional[str] = None
    created_by: str
    created_at: Optional[str] = None
    parsed: Optional[dict[str, Any]] = None
    raw: Optional[str] = None


class AICoachMemoryResponse(BaseModel):
    id: str
    type: str
    key: str
    summary: str
    value: Optional[dict[str, Any]] = None
    confidence: float
    salience: float
    last_reinforced_at: Optional[str] = None
